plutus-core/plutus-ir/src/PlutusIR/Compiler.hs

```diff
@@ -82,11 +88,15 @@ compileToReadable =
 compileReadableToPlc :: Compiling m e uni fun a => Term TyName Name uni fun (Provenance a) -> m (PLCTerm uni fun a)
 compileReadableToPlc =
     NonStrict.compileNonStrictBindings
-    >=> Let.compileLets Let.Types
+    >=> Let.compileLets Let.DataTypes
     >=> Let.compileLets Let.RecTerms
     -- We introduce some non-recursive let bindings while eliminating recursive let-bindings, so we
     -- can eliminate any of them which are unused here.
+    >=> PLC.rename
+    -- There was a bug in renamer handling non-rec terms, so we need to rename
+    -- again.
```

Comment:
I think this should be instead at the start of the simplify pipeline. The point is that `beta` is not safe without global uniqueness, so we should ensure that we have it before we run `beta`. The easiest way to do that is just to put it at the start of the simplifier pipeline. We run `simplifyTerm` in other places... do we have this problem there? Unclear, but we know we're safe if we do `rename` before `beta` always.

Reply:
Also, the comment is good, but let's add a reference to the ticket: https://jira.iohk.io/browse/SCP-2156